cpp/ql/src/experimental/Security/CWE/CWE-758/DangerousUseMallocWithZeroSize.ql

```diff
@@ -0,0 +1,164 @@
+      ab.getLValue().(VariableAccess).getTarget() = fc.getArgument(0).(VariableAccess).getTarget() and
+      ab.getASuccessor*() = fc
+    )
+    or
+    exists(Initializer it |
+      fc.getArgument(0).(VariableAccess).getTarget().getInitializer() = it and
+      (
+        it.getExpr().getAChild*() instanceof AddExpr or
+        it.getExpr().isConstant()
+      )
+    )
+    or
+    exists(Assignment ae |
+      ae.getLValue().(VariableAccess).getTarget() = fc.getArgument(0).(VariableAccess).getTarget() and
+      (
+        ae.getRValue().getAChild*() instanceof AddExpr or
+        ae.getRValue().isConstant()
+      ) and
+      ae.getASuccessor*() = fc
+    )
+    or
+    exists(IfStmt ifs, ComparisonOperation co, Expr ec |
+      ifs.getCondition().getAChild*() = co and
+      co.hasOperands(fc.getArgument(0).(VariableAccess).getTarget().getAnAccess(), ec) and
+      exists(Expr etmp |
+        (
+          etmp.getValue() = "0" or
+          etmp.getValue() = "1"
+        ) and
+        (
+          ec = etmp or
+          globalValueNumber(ec) = globalValueNumber(etmp)
+        )
+      )
+    )
+  )
+}
+
+/** Holds if the function within which the call is found is not called elsewhere. */
+predicate enclosingFunctionNotCall(FunctionCall fc) {
+  exists(Function fn |
+    fn = fc.getEnclosingFunction() and
+    not exists(fn.getACallToThisFunction()) and
+    fn.getAParameter().getAnAccess() = fc.getArgument(0) and
+    not exists(TemplateClass tc | tc.getACanonicalMemberFunction() = fn)
+  )
+}
+
+/** Holds if buffer and length are used in string functions or arrays. */
+predicate dangerousUseBufferAndSize(FunctionCall fc) {
+  exists(Expr e0 |
+    fc.getASuccessor*() = e0 and
+    (
+      e0 instanceof ArrayExpr and
+      globalValueNumber(e0.(ArrayExpr).getArrayBase()) = globalValueNumber(fc) and
+      e0.(ArrayExpr).getArrayOffset().getAChild().(VariableAccess).getTarget() =
+        fc.getArgument(0).(VariableAccess).getTarget()
+      or
+      e0.(FunctionCall).getTarget().hasGlobalOrStdName(["strstr", "strlen", "strcpy", "strcat"]) and
+      (
+        globalValueNumber(e0.(FunctionCall).getAnArgument()) = globalValueNumber(fc)
+        or
+        exists(Assignment aetmp, Assignment ae1tmp |
+          aetmp.getLValue().(VariableAccess).getTarget() =
+            e0.(FunctionCall).getAnArgument().(VariableAccess).getTarget() and
+          ae1tmp.getRValue() = fc and
+          aetmp.getRValue().(VariableAccess).getTarget() =
+            ae1tmp.getLValue().(VariableAccess).getTarget() and
+          aetmp.getASuccessor*() = e0
+        )
+      )
+    )
+  )
+}
+
+/** Holds if this function is `fread` and its argument is `exp`. */
+predicate thisFunctionFread(FunctionCall fc, Expr exp) {
+  fc.getTarget().hasGlobalOrStdName("fread") and
+  globalValueNumber(fc.getArgument(0)) = globalValueNumber(exp)
+}
+
+/** Hods if the length argument is related to the result of the function call. */
```

Comment:
I assume `fc` is the function call mentioned in the QLDoc, but what is the "length argument"? Looking at the code, I _think_ the length argument refers to the first argument of `fc`, but this should be noted in the QLDoc.

Reply:
i fixed it. Have a look, please